{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initial try\n",
    "This notebook contains manual commands for first attempts at establishing the GPAD pipeline.  \n",
    "These steps include an example data set to work with, then manual data processing and analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot\n",
    "import plotly.graph_objs as go\n",
    "init_notebook_mode(connected=True)\n",
    "import ipywidgets as widgets\n",
    "from ipywidgets import interact, interact_manual\n",
    "\n",
    "data_dir = \"../data\"\n",
    "out_dir = \"../output/initial\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data download\n",
    "Downloading the raw data for developing GPAD.  \n",
    "Specifically, tomato assembly, annotation and reads.\n",
    "Start by downloading and extracting the Heinz1706 relevant data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-09-07 18:20:30--  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR157/008/SRR1572628/SRR1572628_1.fastq.gz\n",
      "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.192.7\n",
      "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.192.7|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2212339546 (2.1G) [application/octet-stream]\n",
      "Saving to: ‘../data/SRR1572628_1.fastq.gz’\n",
      "\n",
      "100%[====================================>] 2,212,339,546  527KB/s   in 68m 45s\n",
      "\n",
      "2019-09-07 19:29:16 (524 KB/s) - ‘../data/SRR1572628_1.fastq.gz’ saved [2212339546/2212339546]\n",
      "\n",
      "--2019-09-07 19:29:17--  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR157/008/SRR1572628/SRR1572628_2.fastq.gz\n",
      "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.192.7\n",
      "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.192.7|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2285974731 (2.1G) [application/octet-stream]\n",
      "Saving to: ‘../data/SRR1572628_2.fastq.gz’\n",
      "\n",
      "100%[====================================>] 2,285,974,731  529KB/s   in 71m 23s\n",
      "\n",
      "2019-09-07 20:40:41 (521 KB/s) - ‘../data/SRR1572628_2.fastq.gz’ saved [2285974731/2285974731]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# download Heinz data\n",
    "!wget ftp://ftp.solgenomics.net/tomato_genome/Heinz1706/assembly/build_4.00/S_lycopersicum_chromosomes.4.00.fa -P $data_dir\n",
    "!wget ftp://ftp.solgenomics.net/tomato_genome/Heinz1706/annotation/ITAG4.0_release/ITAG4.0_gene_models.gff  -P $data_dir\n",
    "!wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR157/008/SRR1572628/SRR1572628_1.fastq.gz  -P $data_dir\n",
    "!wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR157/008/SRR1572628/SRR1572628_2.fastq.gz  -P $data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract reads\n",
    "!pigz -d $data_dir/SRR1572628_1.fastq.gz\n",
    "!pigz -d $data_dir/SRR1572628_2.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align reads\n",
    "Align heinz reads to reference genome.  \n",
    "Data is ~x5 coverage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir $out_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4332493.power8.tau.ac.il\n"
     ]
    }
   ],
   "source": [
    "# index heinz ref\n",
    "index_sh = \"%s/index_genome.sh\" % out_dir\n",
    "!qsub $index_sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4336012.power8.tau.ac.il\n"
     ]
    }
   ],
   "source": [
    "# align reads\n",
    "align_sh = \"%s/align_reads.sh\" % out_dir\n",
    "!qsub $align_sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4377619.power8.tau.ac.il\n"
     ]
    }
   ],
   "source": [
    "# convert to bam and sort\n",
    "sort_bam_sh = \"%s/sort_bam.sh\" % out_dir\n",
    "!qsub $sort_bam_sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate gene coverage\n",
    "The goal is to create coverage distributions per gene. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4386592.power8.tau.ac.il\n"
     ]
    }
   ],
   "source": [
    "# start by creating a bedGraph\n",
    "make_bedgraph_sh = \"%s/make_bedGraph.sh\" % out_dir\n",
    "!qsub $make_bedgraph_sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../output/initial/ITAG4.0_gene_models_exons.bed\n",
      "../data/ITAG4.0_gene_models.gff\n"
     ]
    }
   ],
   "source": [
    "# convert gff exons to bed\n",
    "gff_file = \"%s/ITAG4.0_gene_models.gff\" % data_dir\n",
    "exons_bed = \"%s/ITAG4.0_gene_models_exons.bed\" % out_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "grep -v '#' ../data/ITAG4.0_gene_models.gff | awk '$3 == \"exon\" {split($9,a,\";\"); split(a[2],b,\"=\"); split(b[2],c,\":\"); print $1\"\\t\"$4\"\\t\"$5\"\\t\"c[2]}' > ../output/initial/ITAG4.0_gene_models_exons.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4401603.power8.tau.ac.il\n"
     ]
    }
   ],
   "source": [
    "# intersect exons with bedGraph. Use -wao to keep gene names and calculate number of overlapping bp\n",
    "intersect_sh = \"%s/intersect_exons_with_bedgraph.sh\" % out_dir\n",
    "!qsub $intersect_sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# process intersect result\n",
    "exons_cov_file = \"%s/exons_coverage.tsv\" % out_dir\n",
    "headers = [\"exon_chrom\", \"exon_start\", \"exon_end\", \"gene\",\n",
    "           \"cov_block_chrom\", \"cov_block_start\", \"cov_block_end\", \"coverage\", \"overlap\"]\n",
    "exons_cov_df = pd.read_csv(exons_cov_file, names=headers, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_chrom</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>gene</th>\n",
       "      <th>cov_block_chrom</th>\n",
       "      <th>cov_block_start</th>\n",
       "      <th>cov_block_end</th>\n",
       "      <th>coverage</th>\n",
       "      <th>overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93750</td>\n",
       "      <td>94430</td>\n",
       "      <td>Solyc00g500001.1.1</td>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93749</td>\n",
       "      <td>93762</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93750</td>\n",
       "      <td>94430</td>\n",
       "      <td>Solyc00g500001.1.1</td>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93762</td>\n",
       "      <td>93766</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93750</td>\n",
       "      <td>94430</td>\n",
       "      <td>Solyc00g500001.1.1</td>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93766</td>\n",
       "      <td>93770</td>\n",
       "      <td>13</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93750</td>\n",
       "      <td>94430</td>\n",
       "      <td>Solyc00g500001.1.1</td>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93770</td>\n",
       "      <td>93772</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93750</td>\n",
       "      <td>94430</td>\n",
       "      <td>Solyc00g500001.1.1</td>\n",
       "      <td>SL4.0ch00</td>\n",
       "      <td>93772</td>\n",
       "      <td>93786</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  exon_chrom  exon_start  exon_end                gene cov_block_chrom  \\\n",
       "0  SL4.0ch00       93750     94430  Solyc00g500001.1.1       SL4.0ch00   \n",
       "1  SL4.0ch00       93750     94430  Solyc00g500001.1.1       SL4.0ch00   \n",
       "2  SL4.0ch00       93750     94430  Solyc00g500001.1.1       SL4.0ch00   \n",
       "3  SL4.0ch00       93750     94430  Solyc00g500001.1.1       SL4.0ch00   \n",
       "4  SL4.0ch00       93750     94430  Solyc00g500001.1.1       SL4.0ch00   \n",
       "\n",
       "   cov_block_start  cov_block_end  coverage  overlap  \n",
       "0            93749          93762        11       12  \n",
       "1            93762          93766        12        4  \n",
       "2            93766          93770        13        4  \n",
       "3            93770          93772        15        2  \n",
       "4            93772          93786        14       14  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_cov_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov_dist_per_gene = exons_cov_df.groupby(['gene', 'coverage'])['overlap'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene                coverage\n",
       "Solyc00g005280.1.1  13           2\n",
       "                    14          21\n",
       "                    15           8\n",
       "                    16          25\n",
       "                    17          56\n",
       "                    18          49\n",
       "                    19          50\n",
       "                    20          40\n",
       "                    21          29\n",
       "                    22          52\n",
       "                    23           5\n",
       "                    24          10\n",
       "                    25           9\n",
       "                    26           4\n",
       "                    27           1\n",
       "                    28          13\n",
       "                    29          14\n",
       "                    30          16\n",
       "                    31           5\n",
       "Solyc00g006650.1.1  7           11\n",
       "                    8           35\n",
       "                    9           15\n",
       "                    10           2\n",
       "                    11          22\n",
       "                    12           8\n",
       "                    13          20\n",
       "                    14           4\n",
       "                    21           2\n",
       "                    22           6\n",
       "                    23          16\n",
       "Name: overlap, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_dist_per_gene.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "example = cov_dist_per_gene['Solyc00g005280.1.1']\n",
    "example_vec = example.index.repeat(example)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x2b27f2e5e6d8>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEGCAYAAABsLkJ6AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3deXxU5b348c83G4EAgZCwJpAAYd8JYRFwV7RV1IJArWJrS7X1dvHXXrX2Wq+1t8u9rbX3qhXrXhEBtWKLct2oypqwbyIBQhKWkBBIwpJ1vr8/5sQ7xgmZrJPM+b5fr3nlzHOec873cWS+c55zzvOIqmKMMcZ9woIdgDHGmOCwBGCMMS5lCcAYY1zKEoAxxriUJQBjjHGpiGAH0BDx8fGanJwc7DCMMaZd2bx5c6GqJtQub1cJIDk5mczMzGCHYYwx7YqIHPZXbl1AxhjjUpYAjDHGpSwBGGOMS1kCMMYYl7IEYIwxLhVQAhCRWSKyT0SyROQ+P+vvEZE9IrJDRN4XkQE+6xaKyH7ntdCnfKKI7HT2+ScRkeZpkjHGmEDUmwBEJBx4HLgGGAEsEJERtaptBdJUdQywAvids20c8AtgMpAO/EJEujvbPAksAlKd16wmt8YYY0zAAjkDSAeyVPWgqlYAS4HZvhVU9UNVPee83QAkOstXA++qapGqngLeBWaJSB+gq6quV+941C8CNzRDe4wxxgQokATQD8j1eZ/nlNXlDuDterbt5yzXu08RWSQimSKSWVBQEEC4xhhjAhHIk8D++ub9ziIjIt8A0oCL69k24H2q6mJgMUBaWprNXtPGLdmY0+htvz65fzNGYoypTyBnAHlAks/7ROBo7UoicgXwAHC9qpbXs20e/9dNVOc+jTHGtJxAEkAGkCoiKSISBcwHVvpWEJHxwFN4v/xP+KxaDVwlIt2di79XAatV9RhQKiJTnLt/bgPebIb2GGOMCVC9XUCqWiUid+P9Mg8HnlXV3SLyMJCpqiuB/wQ6A8uduzlzVPV6VS0SkV/iTSIAD6tqkbN8F/A80BHvNYO3McYY02oCGg1UVVcBq2qVPeizfMUFtn0WeNZPeSYwKuBIjTHGNCt7EtgYY1zKEoAxxriUJQBjjHEpSwDGGONSlgCMMcalLAEYY4xLWQIwxhiXsgRgjDEuZQnAGGNcyhKAMca4lCUAY4xxKUsAxhjjUpYAjDHGpSwBGGOMS1kCMMYYl7IEYIwxLmUJwBhjXCqgBCAis0Rkn4hkich9ftbPFJEtIlIlInN8yi8VkW0+rzIRucFZ97yIHPJZN675mmWMMaY+9U4JKSLhwOPAlUAekCEiK1V1j0+1HOB24Ce+26rqh8A4Zz9xQBbwvz5VfqqqK5rSAGOMMY0TyJzA6UCWqh4EEJGlwGzg8wSgqtnOOs8F9jMHeFtVzzU6WmOMMc0mkC6gfkCuz/s8p6yh5gOv1Cr7lYjsEJFHRaSDv41EZJGIZIpIZkFBQSMOa4wxxp9AEoD4KdOGHERE+gCjgdU+xfcDw4BJQBxwr79tVXWxqqapalpCQkJDDmuMMeYCAkkAeUCSz/tE4GgDj3Mz8IaqVtYUqOox9SoHnsPb1WSMMaaVBJIAMoBUEUkRkSi8XTkrG3icBdTq/nHOChARAW4AdjVwn8YYY5qg3gSgqlXA3Xi7b/YCy1R1t4g8LCLXA4jIJBHJA+YCT4nI7prtRSQZ7xnEP2vt+mUR2QnsBOKBR5reHGOMMYEK5C4gVHUVsKpW2YM+yxl4u4b8bZuNn4vGqnpZQwI1xhjTvOxJYGOMcSlLAMYY41KWAIwxxqUsARhjjEtZAjDGGJeyBGCMMS5lCcAYY1zKEoAxxriUJQBjjHEpSwDGGONSlgCMMcalLAEYY4xLWQIwxhiXsgRgjDEuZQnAGGNcyhKAMca4lCUAY4xxqYASgIjMEpF9IpIlIvf5WT9TRLaISJWIzKm1rlpEtjmvlT7lKSKyUUT2i8irznzDxhhjWkm9CUBEwoHHgWuAEcACERlRq1oOcDuwxM8uzqvqOOd1vU/5b4FHVTUVOAXc0Yj4jTHGNFIgZwDpQJaqHlTVCmApMNu3gqpmq+oOwBPIQUVEgMuAFU7RC8ANAUdtjDGmyQJJAP2AXJ/3efiZ5P0CokUkU0Q2iEjNl3wP4LSqVtW3TxFZ5GyfWVBQ0IDDGmOMuZCIAOqInzJtwDH6q+pRERkIfCAiO4GSQPepqouBxQBpaWkNOa4xxpgLCOQMIA9I8nmfCBwN9ACqetT5exBYA4wHCoFuIlKTgBq0T2OMMU0XSALIAFKdu3aigPnAynq2AUBEuotIB2c5HrgI2KOqCnwI1NwxtBB4s6HBG2OMabx6E4DTT383sBrYCyxT1d0i8rCIXA8gIpNEJA+YCzwlIrudzYcDmSKyHe8X/m9UdY+z7l7gHhHJwntN4JnmbJgxxpgLC+QaAKq6ClhVq+xBn+UMvN04tbdbB4yuY58H8d5hZIwxJgjsSWBjjHEpSwDGGONSlgCMMcalLAEYY4xLWQIwxhiXsgRgjDEuZQnAGGNcyhKAMca4lCUAY4xxKUsAxhjjUpYAjDHGpSwBGGOMS1kCMMYYl7IEYIwxLmUJwBhjXMoSgDHGuJQlAGOMcamAEoCIzBKRfSKSJSL3+Vk/U0S2iEiViMzxKR8nIutFZLeI7BCReT7rnheRQyKyzXmNa54mGWOMCUS9U0KKSDjwOHAlkAdkiMhKn7l9AXKA24Gf1Nr8HHCbqu4Xkb7AZhFZraqnnfU/VdUVTW2EMcaYhgtkTuB0IMuZwxcRWQrMBj5PAKqa7azz+G6oqp/5LB8VkRNAAnAaY4wxQRVIF1A/INfnfZ5T1iAikg5EAQd8in/ldA09KiId6thukYhkikhmQUFBQw9rjDGmDoEkAPFTpg05iIj0AV4CvqmqNWcJ9wPDgElAHHCvv21VdbGqpqlqWkJCQkMOa4wx5gICSQB5QJLP+0TgaKAHEJGuwD+An6vqhppyVT2mXuXAc3i7mowxxrSSQBJABpAqIikiEgXMB1YGsnOn/hvAi6q6vNa6Ps5fAW4AdjUkcGOMMU1TbwJQ1SrgbmA1sBdYpqq7ReRhEbkeQEQmiUgeMBd4SkR2O5vfDMwEbvdzu+fLIrIT2AnEA480a8uMMcZcUCB3AaGqq4BVtcoe9FnOwNs1VHu7vwJ/rWOflzUoUmOMMc3KngQ2xhiXsgRgjDEuFVAXkDEt7Wx5FesPnCQqIowu0RH0j+tEdGR4sMMyJqRZAjBBc76imo/3F7DzSDEnz1Z8YV10ZBjTBydwzajeXDe2L1ERdrJqTHOzBGBanUeVdVmFfLivgLLKaob06kJachzfmNKfao9SfL6SzYdP8f7eE7y3N58/vv8ZP7gslZsmJBIe5u+5RGNMY1gCMK2qstrD8s157DpSzJBenbl6ZG/6xHYEYEbq/z3pPXtcP/79emXNvgJ+/+4+frpiB69syuH3N48jJT4mWOEbE1LsvNq0mnPlVTzzySF2HSnmmlG9WTg1+fMvf39EhEuH9eStu6fz6LyxZJ04w7WPfcySjTmoNmg0EmOMH5YATKuoqvbw0sbDHD19ngXp/ZmRmoD3IfD6iQg3jk/kf398MWnJ3fnZGzt54G+7qKjy1L+xMaZOlgBMi1NV3tx+lMMnzzFnYiKj+8U2aj+9Y6N5/pvp3HnxIJZszOHWZzZSfK6ymaM1xj0sAZgWt+7ASTYfPsWlQxMYk9itSfsKDxPuu2YYj84by9ac08x/egOFZ8qbKVJj3MUSgGlR+SVlvLPrOMP7dOXy4b2abb83jk/kLwvTOFR4hnlPred4cVmz7dsYt7C7gEyL8ajyxtYjdIgM48bx/QgLsM8/UDOHJPDitybzreczmPvUOpZ8ewpJcZ2a9RiBWrIxp0nbf31y/2aKxJjA2RmAaTEbD54kp+gcXxndh84dWua3RnpKHC9/ezIl56uY++f1ZJ040yLHMSYUWQIwLaL4fCWr9+ST2rMz45Ka1u9fn7FJ3Vi6aApVHg/zF6/ns/zSFj2eMaHCEoBpEe/vzafao8we1y/g2z2bYnifrrz63amEibBg8QZLAsYEwBKAaXYnSsvYfPgUU1LiiIuJarXjDkrozNJFUwgP8yaBfcctCRhzIZYATLN7d08+kRFhXDy0Z6sfe6CTBCLCha8/bUnAmAsJKAGIyCwR2SciWSJyn5/1M0Vki4hUicicWusWish+57XQp3yiiOx09vknaY1+AtPi8k6dY/fREmYMjm+xC7/1GZjQmVe+400CCywJGFOnehOAiIQDjwPXACOABSIyola1HOB2YEmtbeOAXwCTgXTgFyLS3Vn9JLAISHVesxrdCtNmvLc3n05R4UwfHB/UOGqSQKSTBD49XhLUeIxpiwI5A0gHslT1oKpWAEuB2b4VVDVbVXcAtQdnuRp4V1WLVPUU8C4wS0T6AF1Vdb16R/V6EbihqY0xwfXp8RI+yz/DRYPj6dAGJnPxdgdNJTJc+PrTGy0JGFNLIAmgH5Dr8z7PKQtEXdv2c5br3aeILBKRTBHJLCgoCPCwJhgWf3SQqPAwJqfEBTuUz6XEx3whCew6UhzskIxpMwJJAP765gMdi7eubQPep6ouVtU0VU1LSEjwV8W0AUdPn2fltqOkJXenU1TbesC8Jgl0jAxn/uINrD9wMtghGdMmBJIA8oAkn/eJwNEA91/XtnnOcmP2adqgZz85hAIXBbnvvy4p8TGsuGsqvWOjWfjcJlbvPh7skIwJukASQAaQKiIpIhIFzAdWBrj/1cBVItLdufh7FbBaVY8BpSIyxbn75zbgzUbEb9qA0rJKXtmUw3Vj+tC9U+vd999QfWI7svy7UxnRpyt3/XUzyzJz69/ImBBWbwJQ1Srgbrxf5nuBZaq6W0QeFpHrAURkkojkAXOBp0Rkt7NtEfBLvEkkA3jYKQO4C/gLkAUcAN5u1paZVvOPHcc4W1HNbdOSgx1KvbrHRPHytydz0eB4/nXFDp765wGbXcy4VkCdtaq6ClhVq+xBn+UMvtil41vvWeBZP+WZwKiGBGvaplczc0nt2ZnxSd349Fjj77lvrRE1YzpE8MzCSdyzbBu/fvtTjpeU8fOvjLAJ543rtK2rdabd2Z9fytac0zxw7fBWGfOnuURFhPGn+ePp2SWaZ9ce4nhxGY/OG0d0G7h91ZjWYgnANMmyzFwiwoQbJwR6Z3DbERYmPHjdCPp2i+aRf+yl8MxGnr4tjW5t+DqGMc3JxgIyjVZR5eH1LUe4fHhP4jt3CHY4jfbtGQP57wXj2Z5bzJw/ryfv1Llgh2RMq7AEYBrtg09PcPJsBfMmJdVfuY27bmxfXvhWOvklZXztyXVknbDxg0zoswRgGm1ZZi69unZgZmpoPKA3dVAPlt85lWoP3PzUBntq2IQ8SwCmUfJLyliz7wRfm5BIRHjo/G80rHdXlt85leiIMBY8vYEdeaeDHZIxLSZ0/uWaVrVicx4ehZvT2n/3T20p8TEsv2saXaMjWfjsJptdzIQsSwCmwVSV5Zm5pKfEkRwfE+xwWkS/bh1Z8p3JRIaHcctfNnL45Nlgh2RMs7MEYBps06Eisk+eY14I/vr3NaBHDC9/ezJV1R6++VwGxecqgx2SMc3KEoBpsFczc+ncIYJrR/cJdigtLrVXF566NY3cU+e4+5UtVFXXnvLCmPbLEoBpkNKySlbtPMZ1Y/vSMcodT82mp8TxyA2j+Hh/If+x6tNgh2NMs7EEYBrkre3HKKv0hMS9/w0xb1J/bp+WzLNrD9lQ0iZkWAIwDfJqZi5DenVmbGJssENpdT+7djij+nXlvtd2kF9SFuxwjGkySwAmYPuOl7I99zQ3pyW1q4HfmktURBiPzR/P+cpqfrJ8Ox6PDSNt2jdLACZgyzJziQwXbhzf/gZ+ay6DEjrz4FdH8vH+Ql7eeDjY4RjTJJYATEAqqjy8sfUIVwzvRY92PPBbc1iQnsT0wfH87p19nCi1riDTflkCMAF5f28+RWcruNllF3/9ERF+ecMoyqs9PPL3vcEOx5hGCygBiMgsEdknIlkicp+f9R1E5FVn/UYRSXbKbxGRbT4vj4iMc9atcfZZs65nczbMNK9XM3Pp3TU6ZAZ+a6qU+Bi+d8kgVm4/yif7C4MdjjGNUm8CEJFw4HHgGmAEsEBERtSqdgdwSlUHA48CvwVQ1ZdVdZyqjgNuBbJVdZvPdrfUrFfVE83QHtMCjhWf56PPCpgzMdGmTfRx58WDSImP4Rcrd1FtF4RNOxTIjGDpQJaqHgQQkaXAbGCPT53ZwEPO8grgf0RE9IuzbS8AXmlyxKbVveYM/DY3ze+0z21Ga80pXCM6Mpz7rxnGopc2s/nwKdJT4pp0fGNaWyBdQP2AXJ/3eU6Z3zqqWgUUAz1q1ZnHlxPAc073z79JHfcVisgiEckUkcyCgoIAwjXNyeNRlmXmMWVgHAN6hObAb01x5YhepA3ozvt786mosmEiTPsSSALw98Vc+3z3gnVEZDJwTlV3+ay/RVVHAzOc163+Dq6qi1U1TVXTEhKs/7m1bTxURE7ROdc9+RsoEeH+a4dRWl7FJ1l2LcC0L4EkgDzA919/InC0rjoiEgHEAkU+6+dT69e/qh5x/pYCS/B2NZk2ZnlmLl06RDBrZOgP/NZYEwfEMaJPVz7eX8CZ8qpgh2NMwAJJABlAqoikiEgU3i/zlbXqrAQWOstzgA9q+v9FJAyYCyytqSwiESIS7yxHAl8FdmHalJKySlbtOsb149wz8FtjXTWiFxVVHtbaWYBpR+pNAE6f/t3AamAvsExVd4vIwyJyvVPtGaCHiGQB9wC+t4rOBPJqLiI7OgCrRWQHsA04Ajzd5NaYZrVy21FXDvzWGD27RjOqXyzrD57kXIWdBZj2IZC7gFDVVcCqWmUP+iyX4f2V72/bNcCUWmVngYkNjNW0suWZuQzr3YXR/dw38FtjXDqsJzuPFLM26yRXjugV7HCMqZc9CWz8+vR4Cdvzil078Ftj9O4azci+XVl3oJDzFdXBDseYelkCMH4ty8gjKjzM1QO/NcalQ3tSXuVh/UG7FmDaPksA5kvKq6p5Y2seV47oRfeYqGCH06707daRob26sP5gEZU2faRp4ywBmC95Z9dxTp2rtIu/jTQ9NZ6z5VVsyz0d7FCMuSBLAOZLlmzMoX9cJ6YPjg92KO3SwPgY+sZG88n+QjxqYwSZtssSgPmCrBNn2HioiPnpSYTZwG+NIiJMT42n4Ew5n+WXBjscY+pkCcB8wSubcogIE+ZOtO6fphjdrxuxHSNtqGjTplkCMJ8rq6zmtS15XD2yNwld3D3rV1OFhwlTB/bgYOFZjp4+H+xwjPHLEoD53Kqdxzh9rrLBwyIb/yYlxxEVEWaDxJk2yxKAAUBVeX5dNgMTYpg6sPZI3qYxOkaFM2lAd3bkneb0uYpgh2PMl1gCMABsyTnFjrxivjkt2S7+NqNpzp1U6w+cDHIkxnyZJQADwLNrs+kaHcFNE9r2rF/tTfdOUYzsG8um7CLKKm14CNO2WAIwHD19nnd2HWd+en9iOgQ0PqBpgBmp8ZRXecg8fCrYoRjzBZYADC9tOIyqctvUAcEOJSQldu9Eco9OrDtQaJPHmzbFEoDLnSmv4uUNh7lqRG8Su3cKdjgha/rgBE6fq2TPsZJgh2LM5+x83+X+uuEwJWVV3HXJoGCHEnRLNua02L6H9elCj5goPtlfwKi+XW2IbdMm2BmAi5VVVvOXjw8xIzWesUndgh1OSAsT4aLB8eSeOk9O0blgh2MMEGACEJFZIrJPRLJE5D4/6zuIyKvO+o0ikuyUJ4vIeRHZ5rz+7LPNRBHZ6WzzJ7GfRK1ueWYuhWfK+d4lg4MdiitM6N+djpHh9mCYaTPqTQAiEg48DlwDjAAWiMiIWtXuAE6p6mDgUeC3PusOqOo453WnT/mTwCIg1XnNanwzTENVVnv48z8PMqF/N6YMjAt2OK4QFRFGekoce46WUHTWHgwzwRfIGUA6kKWqB1W1AlgKzK5VZzbwgrO8Arj8Qr/oRaQP0FVV16uqAi8CNzQ4etNoyzPzOHL6PN+/dLD1R7eiqQN7ECbC2gN2FmCCL5AE0A/I9Xmf55T5raOqVUAxUDOeQIqIbBWRf4rIDJ/6efXsEwARWSQimSKSWVBQEEC4pj7nK6r543ufMXFAdy4b1jPY4bhK146RjEmMZXP2KZs32ARdIAnA38/D2jcz11XnGNBfVccD9wBLRKRrgPv0FqouVtU0VU1LSEgIIFxTn+fWHeJEaTn3zhpmv/6DYHpqPBXVHjKyi4IdinG5QG4DzQN8B4dPBI7WUSdPRCKAWKDI6d4pB1DVzSJyABji1Pcdc8DfPk0LOH2ugifXHOCyYT1JT/ly339L3gppvPrEdmRQQgzrDhQybXAPIsLsZjwTHIH8n5cBpIpIiohEAfOBlbXqrAQWOstzgA9UVUUkwbmIjIgMxHux96CqHgNKRWSKc63gNuDNZmiPqcfjH2ZxpryKn149NNihuNqM1ARKyqrYbvMGmyCqNwE4ffp3A6uBvcAyVd0tIg+LyPVOtWeAHiKShberp+ZW0ZnADhHZjvfi8J2qWnPeexfwFyALOAC83UxtMnXIOlHKc2uzmTMhkeF9ugY7HFdL7dmZvrHRrNlXYPMGm6AJ6ElgVV0FrKpV9qDPchkw1892rwGv1bHPTGBUQ4I1jaeq/GLlbjpFhXPvNcOCHY7riQiXDO3Jkk057DxSHOxwjEvZUBBtUFP74f3N6PWPncdYm3WSh2ePJL6zTffYFozo25WELh1Ys+8EHo/aPAym1dnVJxcoKavkkb/vZWTfrtwy2Ub8bCvCRLhkSAL5JeW8uzc/2OEYF7IE4AK/+vteTpSW8asbRxNuvzLblDGJ3egRE8Wj736Gx4aKNq3MEkCIW7PvBK9m5rJo5iDG2YBvbU54mHDF8F58eryUldvtTmjTuuwaQAgrPl/Jfa/tJLVnZ350RWqwwzF1GJ0Yy44jxfzh3c+4dnQfoiLsd1mgmnK9zN+1Mrex/9NClKrywBs7KThTzn/OHUt0ZHiwQzJ1CBPhX2cNJafoHK9m2IN4pvVYAghRr2bk8vcdx/jxFanW9dMOXDIkgfTkOB57P4vSsspgh2NcwhJACPosv5SH3trNRYN7cJeN9d8uiAgPfGU4J8+W89h7+4MdjnEJSwAhpryymu+9vIXOHSJ4dN44u+unHRmb1I35k5J4bl02n+WXBjsc4wKWAEKIqrJiSx4HC87wp/nj6dklOtghmQb66dXD6BIdwYNv7kJtiAjTwiwBhJCP9hey+2gJ910zjGmD44MdjmmEuJgofnLVUDYcLOL1LUeCHY4JcZYAQsS+46X87+7jjO4Xy3dmDAx2OKYJFqT3J21Adx56azdHT58PdjgmhFkCCAHHS8pYmpFD79hobprQzyZ5aefCw4Tf3zyWao/y0xXb7Qlh02IsAbRzZ8qreHF9NlERYdw2NZkOEXa/fygY0COGn39lBGuzTvL8uuxgh2NClD0J3I5VVnv464bDnC2v4jszBhLbMTLYIZlmtCA9iff35vPrt/cyJjGWtOQvz+AWasqrqtmRV8ymQ0XsPlpM4ZkKTp2toFNUOHExUQzoEUN6ShyTU+LoYaPaNpklgHZKVXl9Sx45RedYkN6fxO6dgh2SaWYiwh9uHsfsxz/hzr9u4a1/uYg+sR2DHVaL2HWkmKUZOby57SilZVUApMTHkNClAwMTYjhf6aHgTDkbDxXx/LpsRLwPzyXFdWJIry6EWbdno1gCaKc+2HeC7XnFXDWiF6P7xQY7HNNCYjtFsvi2NG58fC13vrSZVxZNoVNU6Pyz3Zpzij+8+xkf7y+kQ0QY14zqzbWj+zApOY7uMVFfql9Z7WHnkWI+2Osd5PDDfQX06tqBa0f1IbVXlyC0oH0L6BqAiMwSkX0ikiUi9/lZ30FEXnXWbxSRZKf8ShHZLCI7nb+X+WyzxtnnNufVs7kaFeq2553m/b0nGJ/UjYuHJAQ7HNPChvTqwqPzxrHzSDHffWkzZZXVwQ6pyU6dreDuJVu48Yl17D5aws+uHcamB67gj/PHc9XI3n6//AEiw8OY0L87P7l6KOvuu4x5aUlUVivPrcvmhXXZFJ+3YTQaot6fEs6k7o8DVwJ5QIaIrFTVPT7V7gBOqepgEZkP/BaYBxQC16nqUREZhXde4X4+293iTA1pApRTdI7XNucxoEcnbhxvd/y4xVUje/Pbr43hpyt2cPeSLTz5jYlEhrfPezg++DSfe1/byelzFVw+rCfTU+PpEBHOP3Yca/C+xiZ1Y2Tfrqw/eJL39ubzp/f3M3tcX8Yk2vhXgQjk/6B0IEtVD6pqBbAUmF2rzmzgBWd5BXC5iIiqblXVmkHOdwPRImJXbhrp1LkKXtpwmK4dI/nG5AFEtNMvANM4c9OS+OXskby39wR3vrSZcxVVwQ6pQc6WV3H/6zv51vOZ9IiJ4s3vT+fy4b2afOdaRHgYM1IT+JfLUonvHMXSjFz+tvUIVR5PM0UeugLpTOwH5Pq8zwMm11VHVatEpBjogfcMoMbXgK2qWu5T9pyIVOOdOP4R9fPsu4gsAhYB9O/v3vG7yyureWn9Yao9Hm6bkkJMh9DpBzaBu3VqMojwizd3seDpjTyzMC3oczwHMiZ/duFZVmzJ49TZCmamxnPF8F5syz3drHHEd+7AopmDeHdPPh/tLyC/tIxbJg+gs/1bqVMgPyH99THU/qK+YB0RGYm3W+i7PutvUdXRwAzndau/g6vqYlVNU9W0hAR39nd7VFmakcuJ0jIWTOpPz642xo+b3TplAH/+xkT2HS/hxifWsjOvONgh1amq2sM7u47x9McHUVW+PWMgs0b1abGz1/AwYdao3sxLS+Lo6fM8uSaLwjPl9W/oUoF8CnlAks/7RKD23HWf1xGRCCAWKHLeJwJvALep6oGaDVT1iPO3FFiCt6vJ+PH2zmPsyy/lq2P62p0OBvBeE3jlO1Oorla+9uQ6Xlyf3eYGjztWfJ4n1hzgo/2FpCV35weXpZISH9Mqxx6b1I3vzBhIeZWHpz46aCjPkzEAAA91SURBVENq1CGQBJABpIpIiohEAfOBlbXqrAQWOstzgA9UVUWkG/AP4H5VXVtTWUQiRCTeWY4EvgrsalpTQtPGQydZe+Ak0wb1YMrAHsEOx7Qh4/t35x8/mMFFg3vw4Ju7uf25DI60gS86jyr/3HeCJz48wJnyKm6bOoAbxyfSoZVnpUvs3olFMwcSESY8/fFBDhWebdXjtwf1JgBVrQLuxnsHz15gmaruFpGHReR6p9ozQA8RyQLuAWpuFb0bGAz8W63bPTsAq0VkB7ANOAI83ZwNCwVZJ87w1vajDO3VhWtH9wl2OKYN6h4TxTMLJ/HQdSPIyC7iqj/8k2c/OURldXAugJ48U87ijw6yek8+w/t04YeXpzKsd9egxALQs0s03505kC7RkTy39hCfHi8JWixtUUBXR1R1FbCqVtmDPstlwFw/2z0CPFLHbicGHqb75JeUsWTTYRK6dGDepCR70tHUKSxMuP2iFC4f3osH/raLh/++hyWbcnjgK8O5ZEhCq9wq7FFl46Ei3tl1jPAw4ea0JMYmxraJ25S7dYpi0cyBvLAum79uOMyciUk2TarDLo+3QcXnK3l+XTaRYWHcNiXZJnR3gUDupKnL1yd7745LiuvEC9+cxHt7T/Crf+zhm89lMCm5Oz++cgjTBrXc/BD5JWW8sfUIOUXnSO3ZmZsmJLa5cak6d4jgjukpvLThMMszc6mq9nz+383NLAG0Md4v/0OUVVbznRkD63wi0hh/RIQrR/Ti4iEJLM3I4fEPs/j60xuZ0L8b354xkKtG9Gq2O3BOlJbx2Hv7eWVTDtGR4cydmMi4pG5t4le/P9GR4dw+LZmXNx7m9a1HGJMYy+0XpQQ7rKCyBNCGlFdV892XMikoLef2aSn07RaaA3+Z5lXX2UNEWBjfu2QwmdlFrD1wku+9vIXYjpFM6N+NCf27fz6aZkN/CWcXnuXF9YdZmpFDRZWH9JQ4LhvWq13cbx8ZHsY3Jg9gaUYuD721h7IqD3dePCjYYQVN2//EXMLjUf7fsu1sOFjEzWmJDO7ZOdghmRAQGR7G1EHxTB7Yg73HSsjILmLNvgI+3FdAn9hohvbuQkp8DKP6daVLtP9uG1Ult+g8H3yaz3t7T7D2QCHhInx1TB9+eMUQ1h842cqtapqI8DAWpPdnU3YRv3n7U85XVPOjK1Lb7JlLS7IE0AaoKr9atZe/7zjG/dcMq/MfYqCa0p9sQlOYCCP7xjKybyzF5yvZkXeavcdK+eizAtbsKwBgQI9O9OoaTUKXDghQXuWh8Ew5WflnKC33DjsxMCGGf7kslVsm96eX80Bie0sA4H1g7I/zxtEhIozH3t9PWWU1910zzHVJwBJAG/Doe/t55pND3D4tmUUzB/LKptz6NzKmkWI7RjIjNYEZqQmcr6hmYM8YduYVs+94KQWl5ew96r1VMioijG6dIrlhfD+G9OrM9NSEVnuQqzWEhwm/+9oYOkaG89RHBzlfWc1D140kLMw9ScASQJA9/mEWf3p/P/PSknjwqyNc9wvEBFfHqHAuHdqTS4e6czT2sDDh4dkjiY4M4+mPD3Guoppf3zS63Y602lCWAIJEVXn8wyz+638/48bx/fiPm0a76peHMW2FiPCza4fTKSqCx97fT0FpOY/fMqFdXNRuKnekuTZGVfnN259+/uX/n3PGEG5f/sYEjYjw4yuH8OubRvNJViHznlrPseLgD6vR0iwBtLKKKg/3vraDpz46yG1TB/D7uWNtXH9j2ogF6f35y21pZBee5br/XktGdlGwQ2pRoX+O04acOlvBXS9vZsPBIn5w2WB+fOUQ6/M3QWd3jX3RpcN68rfvX8SilzazYPEGfv6V4SyclhyS/1btp2cr2XWkmBueWMuWnNP8cd447rlqaEj+D2VMKEjt1YW/ff8iLh6SwENv7eGOFzJDcl4BSwAtzONRFn90gBufWEt5pYeli6Zww/h+9W9ojAmq2I6R/GVhGg9dN4JPsgqZ9cePeHPbkTY370JTWAJoQXuPlTBv8Xr+Y9WnXD6sF+/8aAYT+ncPdljGmACJeEdafevu6fTr3okfLt3Grc9s4rP80mCH1izsGkALOFFSxhNrDvDShsPEdozkP+eMYc7EROvyMaadGtq7C6/fNY0lGw/zu3f2cfUfP+LG8f340eVD6N+jU7DDazRLAM0ou/Asz6/L5pVNOVR5lHmTkvjXq4fSrZON6GlMexceJtw6NZmvjOnLn/95gBfWZfO3rUe4ckQvbp+WwpSBce3uR54lgCY6eaacDz49wetbjrD+4EnCw4SvTejH9y4ZTHIIPTZvjPGKi4niZ9cO547pKbzg/OBbvTufxO4duW5sX64e2ZvR/WLbxbM9EsgFDRGZBTwGhAN/UdXf1FrfAXgR7yxfJ4F5qprtrLsfuAOoBn6gqqsD2ac/aWlpmpmZGXDjmpuqcrykjF1HSth8+BSbDp1ka+5pVCEpriPz0pKYMzGJ3rHRTTqO3ZZnTMtrrglhyiqrWbXzGCu3H+Xj/YVUe5RunSKZNqgH45K6MTaxG4N6dqZHTFTQzhBEZLOqptUur/cMQETCgceBK4E8IENEVqrqHp9qdwCnVHWwiMwHfgvME5EReCeRHwn0Bd4TkSHONvXts9lUe5QqjwePB6pVqfYoHo9S5VE8qlRWezhXUc3Z8qrP/54+V0l+SRn5pWXkl5STX1LGoYKzn4+KGBnuHV3xB5elcuWIXozs27Xdnf4ZY5ouOjKcmyYkctOERIrOVvDx/gI+3l/IhoMnWbXz+Of1YqLC6d8jhgFxnUiK60j3mCi6dYyiW6dIunWMpHN0BFERYUSFhxEZHkZUhO9fISo8rNm/YwLpAkoHslT1IICILAVmA75f1rOBh5zlFcD/iDfS2cBSVS0HDjmTxqc79erbZ7O544WMz4e8bajunSI/HyK3ZlTEYX26MrpfrE3VaIz5griYKGaP68fscd5bvQvPlLMzr5hDhWfJKTrH4ZNn+exEKR/sO0FFladB+37vnpkM7tmlWeMNJAH0A3zHJ84DJtdVR1WrRKQY6OGUb6i1bc1N8PXtEwARWQQsct6eEZF9AcTcUPFAob8Vh1vgYG1Ane0NYdbm0Neg9t7SgoG0hNTf+i0OtM0D/BUGkgD8nXPUvnBQV526yv09f+D3YoSqLgYWXyjAphKRTH/9Y6HKbe0Fa7MbuK290PQ2B/IgWB6Q5PM+EThaVx0RiQBigaILbBvIPo0xxrSgQBJABpAqIikiEoX3ou7KWnVWAgud5TnAB+q9vWglMF9EOohICpAKbApwn8YYY1pQvV1ATp/+3cBqvLdsPququ0XkYSBTVVcCzwAvORd5i/B+oePUW4b34m4V8H1VrQbwt8/mb17AWrSLqQ1yW3vB2uwGbmsvNLHNAT0HYIwxJvTYYHDGGONSlgCMMcalXJUARORZETkhIrt8yuJE5F0R2e/8Danxmuto80MickREtjmva4MZY3MSkSQR+VBE9orIbhH5oVMesp/zBdocyp9ztIhsEpHtTpv/3SlPEZGNzuf8qnOTSbt3gfY+LyKHfD7jcQ3ar5uuAYjITOAM8KKqjnLKfgcUqepvROQ+oLuq3hvMOJtTHW1+CDijqv8VzNhagoj0Afqo6hYR6QJsBm4AbidEP+cLtPlmQvdzFiBGVc+ISCTwCfBD4B7gdVVdKiJ/Brar6pPBjLU5XKC9dwJ/V9UVjdmvq84AVPUjvHcp+ZoNvOAsv4D3H07IqKPNIUtVj6nqFme5FNiL9+nzkP2cL9DmkKVeZ5y3kc5LgcvwDkcDIfQ5X6C9TeKqBFCHXqp6DLz/kICeQY6ntdwtIjucLqKQ6Q7xJSLJwHhgIy75nGu1GUL4cxaRcBHZBpwA3gUOAKdVtcqp4jv0TLtXu72qWvMZ/8r5jB8V78jMAbME4E5PAoOAccAx4PfBDaf5iUhn4DXgR6paEux4WoOfNof056yq1ao6Du9IAunAcH/VWjeqllO7vSIyCrgfGAZMAuKABnVrWgKAfKcPtaYv9USQ42lxqprv/M/kAZ7m/0ZoDQlOH+lrwMuq+rpTHNKfs782h/rnXENVTwNrgClAN2c4GgjRIWZ82jvL6f5TZ8Tl52jgZ2wJ4IvDWCwE3gxiLK2i5ovQcSOwq6667Y1zsewZYK+q/sFnVch+znW1OcQ/5wQR6eYsdwSuwHvt40O8w9FACH3OdbT3U58fNYL3ekeDPmO33QX0CnAJ3iFU84FfAH8DlgH9gRxgrqqGzEXTOtp8Cd5uAQWyge/W9I+3dyIyHfgY2AnUDLj+M7x94iH5OV+gzQsI3c95DN6LvOF4f8guU9WHRWQgsBRvd8hW4BvOr+N27QLt/QBIwDvy8jbgTp+LxfXv100JwBhjzP+xLiBjjHEpSwDGGONSlgCMMcalLAEYY4xLWQIwxhiXsgRgjDEuZQnAmBbg8zSqMW2WJQDjOiJymzN41nYReUlEBojI+07Z+yLSX0RiRSRbRMKcbTqJSK6IRIrIIBF5R0Q2i8jHIjLMqfO8iPxBRD4Efisi6SKyTkS2On+H+uxrmXO8V53x69OcdVeJyHoR2SIiy53xfYxpEfYrxbiKiIwEHgAuUtVCEYnD+4Tli6r6goh8C/iTqt4gItuBi/EOL3AdsFpVK0VkMd4nLveLyGTgCbzDEAMMAa5Q1WoR6QrMVNUqEbkC+A/ga8D3gFOqOsYZ0GubE1s88HNn+7Mici/e8e0fbo3/NsZ9LAEYt7kMWKGqhQCqWiQiU4GbnPUvAb9zll8F5uFNAPOBJ5xf5NOA5d7hVwDwHYJ3uapWO8uxwAsikop3OIZIp3w68Jhz/F0issMpnwKMANY6+44C1jdHo43xxxKAcRuh/iGCa9avBH7tnCVMBD4AYvCOOV/X1HtnfZZ/CXyoqjc64/Sv8YmhrtjeVdUF9cRnTLOwawDGbd4HbhaRHuCdKxhYh/cXPsAteKfbwxlUaxPeX+t/d4ZWLgEOichcZ3sRkbF1HCsWOOIs3+5T/gne6RoRkRHAaKd8A3CRiAx21nUSkSFNa64xdbMEYFxFVXcDvwL+6fTx/wH4AfBNpyvmVrxzrdZ4FfiG87fGLcAdzva78U436c/v8J5BrMU7imONJ4AE53j3AjuAYlUtwJsoXnHWbcA72YcxLcJGAzWmlYlIOBCpqmUiMgjvWckQVa0IcmjGZewagDGtrxPwoTOLlwB32Ze/CQY7AzDGGJeyawDGGONSlgCMMcalLAEYY4xLWQIwxhiXsgRgjDEu9f8Bz9IwVugPEccAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(example_vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# visualize histograms of gene coverage\n",
    "figs = []\n",
    "for gene_name in cov_dist_per_gene.index.get_level_values(0).unique()[1000:1100]:\n",
    "    gene = cov_dist_per_gene[gene_name]\n",
    "    vec = gene.index.repeat(gene)\n",
    "    data = go.Histogram(x=vec, xbins=dict(size=1), histnorm='percent')\n",
    "    layout = go.Layout(\n",
    "        title=gene_name,\n",
    "        xaxis=dict(title='Coverage'),\n",
    "        yaxis=dict(title='% of exon bp')\n",
    "        )\n",
    "    fig = go.Figure(data=data, layout=layout)\n",
    "    figs.append(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "87ac936616c74a458d2ad24dde5df31d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(IntSlider(value=50, description='i'), Output()), _dom_classes=('widget-interact',))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@interact\n",
    "def show_hist_for_gene(i=(0,100,1)):\n",
    "    figs[i].show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/groups/itay_mayrose/liorglic/miniconda3/envs/jupyterlab/lib/python3.7/site-packages/ipykernel_launcher.py:12: RuntimeWarning:\n",
      "\n",
      "invalid value encountered in long_scalars\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# collect summary stats accross all genes\n",
    "gene_names = []\n",
    "means = []\n",
    "medians = []\n",
    "perc_covs = []\n",
    "seq_depth = 5\n",
    "for gene_name in cov_dist_per_gene.index.get_level_values(0).unique():\n",
    "    gene = cov_dist_per_gene[gene_name]\n",
    "    vec = pd.Series(gene.index.repeat(gene))\n",
    "    mean = vec.mean()\n",
    "    median = vec.median()\n",
    "    perc_cov = vec[vec > seq_depth/2].sum()/vec.sum()*100  # % of gene with coverage > sequencing depth/2\n",
    "    gene_names.append(gene_name)\n",
    "    means.append(mean)\n",
    "    medians.append(median)\n",
    "    perc_covs.append(perc_cov)\n",
    "stats_dict = {'gene': gene_names, 'mean_cov': means, 'median_cov': medians, 'perc_cov': perc_covs}\n",
    "gene_coverage_stats_df = pd.DataFrame(stats_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>mean_cov</th>\n",
       "      <th>median_cov</th>\n",
       "      <th>perc_cov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Solyc00g005280.1.1</td>\n",
       "      <td>20.276284</td>\n",
       "      <td>19.0</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Solyc00g006650.1.1</td>\n",
       "      <td>13.176471</td>\n",
       "      <td>11.0</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Solyc00g007330.1.1</td>\n",
       "      <td>3.696731</td>\n",
       "      <td>3.0</td>\n",
       "      <td>90.271424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Solyc00g007340.1.1</td>\n",
       "      <td>5.032609</td>\n",
       "      <td>5.0</td>\n",
       "      <td>97.516199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Solyc00g007350.1.1</td>\n",
       "      <td>5.427602</td>\n",
       "      <td>6.0</td>\n",
       "      <td>98.416007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene   mean_cov  median_cov    perc_cov\n",
       "0  Solyc00g005280.1.1  20.276284        19.0  100.000000\n",
       "1  Solyc00g006650.1.1  13.176471        11.0  100.000000\n",
       "2  Solyc00g007330.1.1   3.696731         3.0   90.271424\n",
       "3  Solyc00g007340.1.1   5.032609         5.0   97.516199\n",
       "4  Solyc00g007350.1.1   5.427602         6.0   98.416007"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_coverage_stats_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "810dd1a91c1b41d8b1786509c42e2ccd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='stat', options=('mean_cov', 'median_cov', 'perc_cov'), value='mean…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@interact\n",
    "def hist_stat(stat=['mean_cov','median_cov','perc_cov']):\n",
    "    data = go.Histogram(x=gene_coverage_stats_df[stat])\n",
    "    layout = go.Layout(\n",
    "        title=\"Histogram of mean coverage\",\n",
    "        xaxis=dict(title=stat),\n",
    "        yaxis=dict(title='# of genes')\n",
    "        )\n",
    "    fig = go.Figure(data=data, layout=layout)\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create negative examples\n",
    "So far all examples are \"positive\" genes, in the sense that they are assumed to be real genes that exist in the ref genome.  \n",
    "Next step is to look at \"negative\" examples - genes that do not exist in the genome, and how their coverage distributions look.  \n",
    "To get a list of negative genes:\n",
    "1. Blast sequences from relative species against the ref genome\n",
    "2. Look for genes that are not found (use some criteria) in the ref\n",
    "3. Align reads from ref to negative genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-09-14 17:42:13--  ftp://ftp.solgenomics.net/genomes/Solanum_pennellii/annotations/Spenn-v2-cds-annot.fa\n",
      "           => ‘../data/Spenn-v2-cds-annot.fa.1’\n",
      "Resolving ftp.solgenomics.net (ftp.solgenomics.net)... 132.236.81.147\n",
      "Connecting to ftp.solgenomics.net (ftp.solgenomics.net)|132.236.81.147|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/Solanum_pennellii/annotations ... done.\n",
      "==> SIZE Spenn-v2-cds-annot.fa ... 78585655\n",
      "==> PASV ... done.    ==> RETR Spenn-v2-cds-annot.fa ... done.\n",
      "Length: 78585655 (75M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 78,585,655  19.3MB/s   in 4.7s   \n",
      "\n",
      "2019-09-14 17:42:21 (16.0 MB/s) - ‘../data/Spenn-v2-cds-annot.fa.1’ saved [78585655]\n",
      "\n",
      "--2019-09-14 17:42:21--  ftp://ftp.solgenomics.net/genomes/Solanum_pimpinellifolium/LA0480/annotation/s.pimp.maker.transcripts.FINAL.UTR.fna\n",
      "           => ‘../data/s.pimp.maker.transcripts.FINAL.UTR.fna’\n",
      "Resolving ftp.solgenomics.net (ftp.solgenomics.net)... 132.236.81.147\n",
      "Connecting to ftp.solgenomics.net (ftp.solgenomics.net)|132.236.81.147|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/Solanum_pimpinellifolium/LA0480/annotation ... done.\n",
      "==> SIZE s.pimp.maker.transcripts.FINAL.UTR.fna ... 38168721\n",
      "==> PASV ... done.    ==> RETR s.pimp.maker.transcripts.FINAL.UTR.fna ... done.\n",
      "Length: 38168721 (36M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 38,168,721  12.6MB/s   in 2.9s   \n",
      "\n",
      "2019-09-14 17:42:26 (12.6 MB/s) - ‘../data/s.pimp.maker.transcripts.FINAL.UTR.fna’ saved [38168721]\n",
      "\n",
      "--2019-09-14 17:42:26--  ftp://ftp.solgenomics.net/genomes/Solanum_chilense/Stam_et_al_2019/Solanum_chilense.cds.fa\n",
      "           => ‘../data/Solanum_chilense.cds.fa.1’\n",
      "Resolving ftp.solgenomics.net (ftp.solgenomics.net)... 132.236.81.147\n",
      "Connecting to ftp.solgenomics.net (ftp.solgenomics.net)|132.236.81.147|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/Solanum_chilense/Stam_et_al_2019 ... done.\n",
      "==> SIZE Solanum_chilense.cds.fa ... 30765896\n",
      "==> PASV ... done.    ==> RETR Solanum_chilense.cds.fa ... done.\n",
      "Length: 30765896 (29M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 30,765,896  11.7MB/s   in 2.5s   \n",
      "\n",
      "2019-09-14 17:42:31 (11.7 MB/s) - ‘../data/Solanum_chilense.cds.fa.1’ saved [30765896]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# download transcripts data from wild relatives\n",
    "#species include: S. pennellii, S. pimpinellifolium and S. chilense, \n",
    "!wget ftp://ftp.solgenomics.net/genomes/Solanum_pennellii/annotations/Spenn-v2-cds-annot.fa -P $data_dir\n",
    "!wget ftp://ftp.solgenomics.net/genomes/Solanum_pimpinellifolium/LA0480/annotation/s.pimp.maker.transcripts.FINAL.UTR.fna  -P $data_dir\n",
    "!wget ftp://ftp.solgenomics.net/genomes/Solanum_chilense/Stam_et_al_2019/Solanum_chilense.cds.fa -P $data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concat all transcripts into one file\n",
    "relative_transcripts_fasta = \"%s/relative_transcripts.fasta\" % out_dir\n",
    "files = \"%s/Spenn-v2-cds-annot.fa %s/s.pimp.maker.transcripts.FINAL.UTR.fna %s/Solanum_chilense.cds.fa\" %(data_dir,data_dir,data_dir)\n",
    "!cat $files > $relative_transcripts_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5163978.power8.tau.ac.il\r\n"
     ]
    }
   ],
   "source": [
    "# use minimap2 to map genes from relative species to ref genome\n",
    "minimap_sh = \"%s/minimap.sh\" % out_dir\n",
    "!qsub $minimap_sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To simulate a realistic scenario, align Heinz reads to each relative species genome, then repeat the procedure above to get coverage stats for their genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-09-14 17:20:41--  ftp://ftp.solgenomics.net/genomes/Solanum_pennellii/Spenn.fasta\n",
      "           => ‘../output/initial/Spenn.fasta’\n",
      "Resolving ftp.solgenomics.net (ftp.solgenomics.net)... 132.236.81.147\n",
      "Connecting to ftp.solgenomics.net (ftp.solgenomics.net)|132.236.81.147|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/Solanum_pennellii ... done.\n",
      "==> SIZE Spenn.fasta ... 1003809056\n",
      "==> PASV ... done.    ==> RETR Spenn.fasta ... done.\n",
      "Length: 1003809056 (957M) (unauthoritative)\n",
      "\n",
      "100%[====================================>] 1,003,809,056 18.9MB/s   in 53s    \n",
      "\n",
      "2019-09-14 17:21:37 (18.2 MB/s) - ‘../output/initial/Spenn.fasta’ saved [1003809056]\n",
      "\n",
      "--2019-09-14 17:21:38--  ftp://ftp.solgenomics.net/genomes/Solanum_pimpinellifolium/LA0480/assembly/tomato-scaffolds.abyss.77.fasta.gz\n",
      "           => ‘../output/initial/tomato-scaffolds.abyss.77.fasta.gz’\n",
      "Resolving ftp.solgenomics.net (ftp.solgenomics.net)... 132.236.81.147\n",
      "Connecting to ftp.solgenomics.net (ftp.solgenomics.net)|132.236.81.147|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/Solanum_pimpinellifolium/LA0480/assembly ... done.\n",
      "==> SIZE tomato-scaffolds.abyss.77.fasta.gz ... 223789373\n",
      "==> PASV ... done.    ==> RETR tomato-scaffolds.abyss.77.fasta.gz ... done.\n",
      "Length: 223789373 (213M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 223,789,373 19.4MB/s   in 12s    \n",
      "\n",
      "2019-09-14 17:21:54 (17.4 MB/s) - ‘../output/initial/tomato-scaffolds.abyss.77.fasta.gz’ saved [223789373]\n",
      "\n",
      "--2019-09-14 17:22:06--  ftp://ftp.solgenomics.net/genomes/Solanum_chilense/Stam_et_al_2019/Solanum_chilense.scaffolds.fa\n",
      "           => ‘../output/initial/Solanum_chilense.scaffolds.fa’\n",
      "Resolving ftp.solgenomics.net (ftp.solgenomics.net)... 132.236.81.147\n",
      "Connecting to ftp.solgenomics.net (ftp.solgenomics.net)|132.236.81.147|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/Solanum_chilense/Stam_et_al_2019 ... done.\n",
      "==> SIZE Solanum_chilense.scaffolds.fa ... 930361482\n",
      "==> PASV ... done.    ==> RETR Solanum_chilense.scaffolds.fa ... done.\n",
      "Length: 930361482 (887M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 930,361,482 19.6MB/s   in 54s    \n",
      "\n",
      "2019-09-14 17:23:05 (16.5 MB/s) - ‘../output/initial/Solanum_chilense.scaffolds.fa’ saved [930361482]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# download genomes\n",
    "!wget ftp://ftp.solgenomics.net/genomes/Solanum_pennellii/Spenn.fasta -P $data_dir\n",
    "!wget ftp://ftp.solgenomics.net/genomes/Solanum_pimpinellifolium/LA0480/assembly/tomato-scaffolds.abyss.77.fasta.gz -O $data_dir/Spimp.fasta\n",
    "!pigz -d $data_dir/tomato-scaffolds.abyss.77.fasta.gz\n",
    "!wget ftp://ftp.solgenomics.net/genomes/Solanum_chilense/Stam_et_al_2019/Solanum_chilense.scaffolds.fa -P $data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-09-21 11:58:32--  ftp://ftp.solgenomics.net/genomes/Solanum_pimpinellifolium/LA0480/annotation/s.pimp.mgm.FINAL.gff\n",
      "           => ‘../data/s.pimp.mgm.FINAL.gff’\n",
      "Resolving ftp.solgenomics.net (ftp.solgenomics.net)... 132.236.81.147\n",
      "Connecting to ftp.solgenomics.net (ftp.solgenomics.net)|132.236.81.147|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/Solanum_pimpinellifolium/LA0480/annotation ... done.\n",
      "==> SIZE s.pimp.mgm.FINAL.gff ... 30221923\n",
      "==> PASV ... done.    ==> RETR s.pimp.mgm.FINAL.gff ... done.\n",
      "Length: 30221923 (29M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 30,221,923  10.4MB/s   in 2.8s   \n",
      "\n",
      "2019-09-21 11:58:38 (10.4 MB/s) - ‘../data/s.pimp.mgm.FINAL.gff’ saved [30221923]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# download annotations\n",
    "!wget ftp://ftp.solgenomics.net/genomes/Solanum_pimpinellifolium/LA0480/annotation/s.pimp.mgm.FINAL.gff -P $data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# index each genome towards BWA run\n",
    "index_relatives_sh = $out_dir/index_relative_genomes.sh\n",
    "!qsub $index_relatives_sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Align Heinz reads to relatives' genomes\n",
    "align_to_spenn_sh = \"%s/align_reads_to_Spenn.sh\" % out_dir\n",
    "align_to_spimp_sh = \"%s/align_reads_to_Spimp.sh\" % out_dir\n",
    "align_to_schil_sh = \"%s/align_reads_to_Schil.sh\" % out_dir\n",
    "!qsub $align_to_spenn_sh\n",
    "!qsub $align_to_spimp_sh\n",
    "!qsub $align_to_schil_sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make bedgraph for each species\n",
    "make_relatives_bedgraph_sh = \"%s/make_relatives_bedGraphs.sh\" % out_dir\n",
    "!qsub $make_relatives_bedgraph_sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### S. pimp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert gff to bed, only including exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "grep -v '#' ../data//s.pimp.mgm.FINAL.gff | awk '$3 == \"exon\" {split($9,a,\"=\"); split(a[2],b,\":\"); print $1\"\\t\"$4\"\\t\"$5\"\\t\"b[1]}' > ../output/initial/s.pimp.mgm.FINAL_exons.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6254948.power8.tau.ac.il\r\n"
     ]
    }
   ],
   "source": [
    "# intersect exons bed with bedGraph. Use -wao to keep gene names and calculate number of overlapping bp\n",
    "intersect_sh = \"%s/intersect_S_pimp_exons_with_bedgraph.sh\" % out_dir\n",
    "!qsub $intersect_sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/groups/itay_mayrose/liorglic/miniconda3/envs/jupyterlab/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning:\n",
      "\n",
      "Columns (4,7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# process intersect result\n",
    "spimp_exons_cov_file = \"%s/Spimp_exons_coverage.tsv\" % out_dir\n",
    "headers = [\"exon_chrom\", \"exon_start\", \"exon_end\", \"gene\",\n",
    "           \"cov_block_chrom\", \"cov_block_start\", \"cov_block_end\", \"coverage\", \"overlap\"]\n",
    "spimp_exons_cov_df = pd.read_csv(spimp_exons_cov_file, names=headers, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_chrom</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>gene</th>\n",
       "      <th>cov_block_chrom</th>\n",
       "      <th>cov_block_start</th>\n",
       "      <th>cov_block_end</th>\n",
       "      <th>coverage</th>\n",
       "      <th>overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3066150</td>\n",
       "      <td>9692</td>\n",
       "      <td>9951</td>\n",
       "      <td>SPi14667.1</td>\n",
       "      <td>3066150</td>\n",
       "      <td>9658</td>\n",
       "      <td>9695</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3066150</td>\n",
       "      <td>9692</td>\n",
       "      <td>9951</td>\n",
       "      <td>SPi14667.1</td>\n",
       "      <td>3066150</td>\n",
       "      <td>9695</td>\n",
       "      <td>9705</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3066150</td>\n",
       "      <td>9692</td>\n",
       "      <td>9951</td>\n",
       "      <td>SPi14667.1</td>\n",
       "      <td>3066150</td>\n",
       "      <td>9705</td>\n",
       "      <td>9719</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3066150</td>\n",
       "      <td>9692</td>\n",
       "      <td>9951</td>\n",
       "      <td>SPi14667.1</td>\n",
       "      <td>3066150</td>\n",
       "      <td>9719</td>\n",
       "      <td>9739</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3066150</td>\n",
       "      <td>9692</td>\n",
       "      <td>9951</td>\n",
       "      <td>SPi14667.1</td>\n",
       "      <td>3066150</td>\n",
       "      <td>9739</td>\n",
       "      <td>9782</td>\n",
       "      <td>1</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   exon_chrom  exon_start  exon_end        gene cov_block_chrom  \\\n",
       "0     3066150        9692      9951  SPi14667.1         3066150   \n",
       "1     3066150        9692      9951  SPi14667.1         3066150   \n",
       "2     3066150        9692      9951  SPi14667.1         3066150   \n",
       "3     3066150        9692      9951  SPi14667.1         3066150   \n",
       "4     3066150        9692      9951  SPi14667.1         3066150   \n",
       "\n",
       "   cov_block_start  cov_block_end coverage  overlap  \n",
       "0             9658           9695        5        3  \n",
       "1             9695           9705        2       10  \n",
       "2             9705           9719        1       14  \n",
       "3             9719           9739        0       20  \n",
       "4             9739           9782        1       43  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spimp_exons_cov_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "spimp_cov_dist_per_gene = spimp_exons_cov_df.groupby(['gene', 'coverage'])['overlap'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene        coverage\n",
       "SPi00001.1  0           519\n",
       "            1            42\n",
       "            2            42\n",
       "            3            90\n",
       "            4            84\n",
       "            5            62\n",
       "SPi00002.1  1            11\n",
       "            10           28\n",
       "            11            1\n",
       "            2            92\n",
       "            3           210\n",
       "            4           197\n",
       "            5           262\n",
       "            6           240\n",
       "            7           210\n",
       "            8           189\n",
       "            9            83\n",
       "SPi00003.1  0           189\n",
       "            1           120\n",
       "            2           172\n",
       "Name: overlap, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spimp_cov_dist_per_gene.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get list of S. pimp genes not found in Heinz (according to minimap alignment)\n",
    "!samtools view -f 4 $out_dir/relative_transcripts_vs_heinz.sam | grep SPi | cut -f1 > $out_dir/Spimp_genes_not_in_heinz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "spimp_genes_not_in_heinz = pd.read_csv(\"%s/Spimp_genes_not_in_heinz\" % out_dir, names=[\"gene\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract exon cov data for these missing genes\n",
    "spimp_cov_dist_per_gene_df = spimp_cov_dist_per_gene.to_frame()\n",
    "spimp_cov_dist_per_gene_df = spimp_cov_dist_per_gene_df.reset_index(level=['coverage'])\n",
    "spimp_not_in_heinz_cov_dist = spimp_cov_dist_per_gene_df.merge(spimp_genes_not_in_heinz, how='inner', on='gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>coverage</th>\n",
       "      <th>overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPi00467.1</td>\n",
       "      <td>0</td>\n",
       "      <td>1489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPi00473.1</td>\n",
       "      <td>0</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPi05397.1</td>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPi05397.1</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPi05398.1</td>\n",
       "      <td>0</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene coverage  overlap\n",
       "0  SPi00467.1        0     1489\n",
       "1  SPi00473.1        0      371\n",
       "2  SPi05397.1        0      352\n",
       "3  SPi05397.1        1       18\n",
       "4  SPi05398.1        0      266"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spimp_not_in_heinz_cov_dist.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "# visualize histograms of gene coverage\n",
    "figs = []\n",
    "for gene_name in spimp_not_in_heinz_cov_dist['gene'].unique():\n",
    "    gene = spimp_not_in_heinz_cov_dist.loc[spimp_not_in_heinz_cov_dist['gene'] == gene_name]\n",
    "    vec = gene['coverage'].repeat(gene['overlap'])\n",
    "    data = go.Histogram(x=vec, xbins=dict(size=1), histnorm='percent')\n",
    "    layout = go.Layout(\n",
    "        title=gene_name,\n",
    "        xaxis=dict(title='Coverage'),\n",
    "        yaxis=dict(title='% of exon bp')\n",
    "        )\n",
    "    fig = go.Figure(data=data, layout=layout)\n",
    "    figs.append(fig)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cf9d113fad6f40c9878d7900129bab56",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(IntSlider(value=25, description='i', max=50), Output()), _dom_classes=('widget-interact'…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@interact\n",
    "def show_hist_for_gene(i=(0,50,1)):\n",
    "    figs[i].show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coverage</th>\n",
       "      <th>overlap</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SPi00255.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi01527.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi01729.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi01749.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi01871.2</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi02096.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi02519.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi02594.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi02834.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi03027.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi03179.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi03276.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi03370.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi03419.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi03631.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04000.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04018.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04027.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04343.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04484.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04485.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04620.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04728.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi04960.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi05407.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi05543.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi06005.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi06203.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi06558.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi06674.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi17652.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi17821.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi17876.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi17907.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi18187.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi18291.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi18460.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi18530.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi18664.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi20888.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi21038.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi21130.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi21368.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi21582.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi21913.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi21913.2</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi21946.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi22172.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi22531.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi23041.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi23205.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi23498.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi23537.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi23656.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi24136.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi24164.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi24593.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi24919.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi25077.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPi25108.1</th>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           coverage  overlap\n",
       "gene                        \n",
       "SPi00255.1        .        0\n",
       "SPi01527.1        .        0\n",
       "SPi01729.1        .        0\n",
       "SPi01749.1        .        0\n",
       "SPi01871.2        .        0\n",
       "SPi02096.1        .        0\n",
       "SPi02519.1        .        0\n",
       "SPi02594.1        .        0\n",
       "SPi02834.1        .        0\n",
       "SPi03027.1        .        0\n",
       "SPi03179.1        .        0\n",
       "SPi03276.1        .        0\n",
       "SPi03370.1        .        0\n",
       "SPi03419.1        .        0\n",
       "SPi03631.1        .        0\n",
       "SPi04000.1        .        0\n",
       "SPi04018.1        .        0\n",
       "SPi04027.1        .        0\n",
       "SPi04343.1        .        0\n",
       "SPi04484.1        .        0\n",
       "SPi04485.1        .        0\n",
       "SPi04620.1        .        0\n",
       "SPi04728.1        .        0\n",
       "SPi04960.1        .        0\n",
       "SPi05407.1        .        0\n",
       "SPi05543.1        .        0\n",
       "SPi06005.1        .        0\n",
       "SPi06203.1        .        0\n",
       "SPi06558.1        .        0\n",
       "SPi06674.1        .        0\n",
       "...             ...      ...\n",
       "SPi17652.1        .        0\n",
       "SPi17821.1        .        0\n",
       "SPi17876.1        .        0\n",
       "SPi17907.1        .        0\n",
       "SPi18187.1        .        0\n",
       "SPi18291.1        .        0\n",
       "SPi18460.1        .        0\n",
       "SPi18530.1        .        0\n",
       "SPi18664.1        .        0\n",
       "SPi20888.1        .        0\n",
       "SPi21038.1        .        0\n",
       "SPi21130.1        .        0\n",
       "SPi21368.1        .        0\n",
       "SPi21582.1        .        0\n",
       "SPi21913.1        .        0\n",
       "SPi21913.2        .        0\n",
       "SPi21946.1        .        0\n",
       "SPi22172.1        .        0\n",
       "SPi22531.1        .        0\n",
       "SPi23041.1        .        0\n",
       "SPi23205.1        .        0\n",
       "SPi23498.1        .        0\n",
       "SPi23537.1        .        0\n",
       "SPi23656.1        .        0\n",
       "SPi24136.1        .        0\n",
       "SPi24164.1        .        0\n",
       "SPi24593.1        .        0\n",
       "SPi24919.1        .        0\n",
       "SPi25077.1        .        0\n",
       "SPi25108.1        .        0\n",
       "\n",
       "[112 rows x 2 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00001.1        0      519\n",
      "SPi00001.1        1       42\n",
      "SPi00001.1        2       42\n",
      "SPi00001.1        3       90\n",
      "SPi00001.1        4       84\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00002.1        1       11\n",
      "SPi00002.1       10       28\n",
      "SPi00002.1       11        1\n",
      "SPi00002.1        2       92\n",
      "SPi00002.1        3      210\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00003.1        0      189\n",
      "SPi00003.1        1      120\n",
      "SPi00003.1        2      172\n",
      "SPi00003.1        3      239\n",
      "SPi00003.1        4      155\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00004.1        1       29\n",
      "SPi00004.1        2       57\n",
      "SPi00004.1        3       53\n",
      "SPi00004.1        4       82\n",
      "SPi00004.1        5      109\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00005.1        0       30\n",
      "SPi00005.1        1       57\n",
      "SPi00005.1       10      101\n",
      "SPi00005.1       11      142\n",
      "SPi00005.1       12       75\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00006.1        0       29\n",
      "SPi00006.1        1      137\n",
      "SPi00006.1       10      130\n",
      "SPi00006.1       11       50\n",
      "SPi00006.1       12       49\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00007.1        0       97\n",
      "SPi00007.1        1       68\n",
      "SPi00007.1       10      115\n",
      "SPi00007.1       11      142\n",
      "SPi00007.1       12       25\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00008.1        0       49\n",
      "SPi00008.1        1      140\n",
      "SPi00008.1       10      109\n",
      "SPi00008.1       11      116\n",
      "SPi00008.1       12       48\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00009.1        0       61\n",
      "SPi00009.1        1       93\n",
      "SPi00009.1       10       66\n",
      "SPi00009.1       11       58\n",
      "SPi00009.1       12        2\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00010.1        0        4\n",
      "SPi00010.1        1       10\n",
      "SPi00010.1       10       93\n",
      "SPi00010.1       11       54\n",
      "SPi00010.1       12       30\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00011.1        0       16\n",
      "SPi00011.1        1       24\n",
      "SPi00011.1       10      108\n",
      "SPi00011.1        2      111\n",
      "SPi00011.1        3      165\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00012.1        0       25\n",
      "SPi00012.1        1       93\n",
      "SPi00012.1       10      105\n",
      "SPi00012.1       11      126\n",
      "SPi00012.1       12      140\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00013.1        0       17\n",
      "SPi00013.1        1      112\n",
      "SPi00013.1       10       74\n",
      "SPi00013.1       11       32\n",
      "SPi00013.1       12       44\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00014.1        1       21\n",
      "SPi00014.1       10       96\n",
      "SPi00014.1       11       39\n",
      "SPi00014.1       12       54\n",
      "SPi00014.1       13       44\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00015.1        0       21\n",
      "SPi00015.1        1       74\n",
      "SPi00015.1       10       17\n",
      "SPi00015.1       11       42\n",
      "SPi00015.1        2      225\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00016.1        0        2\n",
      "SPi00016.1        1       30\n",
      "SPi00016.1       10       70\n",
      "SPi00016.1       11       25\n",
      "SPi00016.1       12        7\n",
      "           coverage  overlap\n",
      "gene                        \n",
      "SPi00017.1        0      149\n",
      "SPi00017.1        1       58\n",
      "SPi00017.1       10       50\n",
      "SPi00017.1       11       44\n",
      "SPi00017.1       12       29\n",
      "coverage     0\n",
      "overlap     73\n",
      "Name: SPi00018.1, dtype: object\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'repeat'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-116-105c6a682f13>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mgene\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspimp_in_heinz_cov_dist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene_name\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0mvec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'coverage'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrepeat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'overlap'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m     \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mHistogram\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvec\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mxbins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhistnorm\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'percent'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     layout = go.Layout(\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'repeat'"
     ]
    }
   ],
   "source": [
    "# visualize histograms of gene coverage\n",
    "figs = []\n",
    "for gene_name in spimp_in_heinz_cov_dist.index.unique()[0:100]:\n",
    "    gene = spimp_in_heinz_cov_dist.loc[gene_name]\n",
    "    print(gene.head())\n",
    "    vec = gene['coverage'].repeat(gene['overlap'])\n",
    "    data = go.Histogram(x=vec, xbins=dict(size=1), histnorm='percent')\n",
    "    layout = go.Layout(\n",
    "        title=gene_name,\n",
    "        xaxis=dict(title='Coverage'),\n",
    "        yaxis=dict(title='% of exon bp')\n",
    "        )\n",
    "    fig = go.Figure(data=data, layout=layout)\n",
    "    figs.append(fig)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
